tapdb/sqlc/migrations/000020_asset_unique_key.up.sql

```diff
@@ -1,4 +1,107 @@
+-- Step 1: If the assets were spent, some of the duplicates might not have been
+-- updated on that flag. To make sure we can properly group on the spent flag
+-- below, we now update all assets that are spent.
+UPDATE assets
+SET spent = true
+WHERE asset_id IN (SELECT a.asset_id
+                   FROM assets a
+                            JOIN managed_utxos mu
+                                 ON a.anchor_utxo_id = mu.utxo_id
+                            JOIN chain_txns ct
+                                 ON mu.txn_id = ct.txn_id
+                            LEFT JOIN asset_transfer_inputs ati
+                                      ON ati.anchor_point = mu.outpoint
+                   WHERE a.spent = false
+                     AND ati.input_id IS NOT NULL);
+
+-- Step 2: Create a temporary table to store the minimum asset_id for each
+-- unique combination.
+CREATE TABLE tmp_min_assets AS
+SELECT MIN(asset_id) AS min_asset_id,
+       genesis_id,
+       script_key_id,
+       amount,
+       anchor_utxo_id,
+       spent
+FROM assets
+GROUP BY genesis_id, script_key_id, amount, anchor_utxo_id, spent;
+
+-- Step 3: Create a mapping table to track old and new asset_ids.
+CREATE TABLE tmp_asset_id_mapping AS
+SELECT a.asset_id       AS old_asset_id,
+       tmp.min_asset_id AS new_asset_id
+FROM assets a
+         JOIN tmp_min_assets tmp
+              ON a.genesis_id = tmp.genesis_id
+                  AND a.script_key_id = tmp.script_key_id
+                  AND a.amount = tmp.amount
+                  AND a.anchor_utxo_id = tmp.anchor_utxo_id
+                  AND a.spent = tmp.spent;
+
+-- Step 4: To make the next step possible, we need to disable a unique index on
+-- the asset_witnesses table. We'll re-create it later.
+DROP INDEX IF EXISTS asset_witnesses_asset_id_witness_index_unique;
+
+-- Step 5: Update the asset_witnesses and asset_proofs tables to reference the
+-- new asset_ids. But skip any records that already reference the new asset_id.
```

Comment:
I think I would be more confident in this step if we didn't do "skip any records that already reference the new asset_id", which I think is unnecessary. Can't we just set again, value shouldn't change.

Reply:
Hmm, you're right. Since we have disabled/removed the unique constraint on the witness table, we can remove the `AND NOT EXISTS` for that table, since it doesn't matter if we create any collisions.

Unfortunately the `asset_proofs` table has an anonymous/inline `UNIQUE` constraint that we cannot disable or remove. So we need to keep that `AND NOT EXISTS` to skip re-assigning proofs if the asset we're going to keep already has one. Otherwise we run into a unique key constraint violation error.

Reply:
For the proofs, we have the `AND NOT EXISTS`. So it only moves proofs for assets where there is no proof already for `new_asset_id`. If I remove the `AND NOT EXISTS`, I get a `UNIQUE constraint failed: asset_proofs.asset_id` error.